maro/rl/experience/sampler.py

```diff
@@ -2,15 +2,131 @@
     def __init__(self, experience_manager: ExperienceManager):
         super().__init__()
         self.experience_manager = experience_manager
 
     @abstractmethod
     def get(self) -> ExperienceSet:
+        """Sampling logic is defined here."""
         raise NotImplementedError
+
+    def on_put(self, experience_set: ExperienceSet, indexes: List[int]):
+        """Callback to be executed after calling experience_manager.put()."""
+        pass
+
+
+class PrioritizedSampler(AbsSampler):
+    """Sampler for Prioritized Experience Replay (PER).
+
+    References:
+        https://arxiv.org/pdf/1511.05952.pdf
+        https://github.com/rlcode/per
+
+    Args:
+        experience_manager (ExperienceManager): experience manager the sampler is associated with.
+        batch_size (int): mini-batch size. Defaults to 32.
+        alpha (float): Prioritization strength. Sampling probabilities are calculated according to
+            P = p_i^alpha / sum(p_k^alpha). Defaults to 0.6.
+        beta (float): Bias annealing strength using weighted importance sampling (IS) techniques.
+            IS weights are calculated according to (N * P)^(-beta), where P is the sampling probability.
+            This value of ``beta`` should not exceed 1.0, which corresponds to full annealing. Defaults to 0.4.
+        beta_step (float): The amount ``beta`` is incremented by after each get() call until it reaches 1.0.
+            Defaults to 0.001.
+    """
+    def __init__(
+        self,
+        experience_manager: ExperienceManager,
+        batch_size: int = 32,
+        alpha: float = 0.6,
+        beta: float = 0.4,
+        beta_step: float = 0.001
+    ):
+        if beta > 1.0:
+            raise ValueError("beta should be between 0.0 and 1.0")
+        super().__init__(experience_manager)
+        self._sum_tree = np.zeros(2 * self.experience_manager.capacity - 1)
+        self.batch_size = batch_size
+        self.alpha = alpha
+        self.beta = beta
+        self.beta_step = beta_step
+        self.eps = 1e-7
+        self._max_priority = 1e8
+
+    def total(self):
+        """Return the sum of priorities over all experiences."""
+        return self._sum_tree[0]
+
+    def on_put(self, experience_set: ExperienceSet, indexes: List[int]):
+        """Set the priorities of newly added experiences to the maximum value."""
+        self.update(indexes, [self._max_priority] * len(indexes))
+
+    def update(self, indexes, td_errors):
+        """Update priority values at given indexes."""
+        for idx, err in zip(indexes, td_errors):
+            priority = self._get_priority(err)
+            tree_idx = idx + self.experience_manager.capacity - 1
+            delta = priority - self._sum_tree[tree_idx]
+            self._sum_tree[tree_idx] = priority
+            self._update(tree_idx, delta)
+
+    def get(self):
+        """Priority-based sampling."""
+        indexes, priorities = [], []
+        segment_len = self.total() / self.batch_size
```

Comment:
why split data into segments? to avoid duplicated samples?

Reply:
I believe this is called stratified sampling (see the last paragraph on Page 4 of the paper) to achieve minibatch balancing.